asyncpp: add new library to conan

petiaccja opened the pull request:
Library name and version:  **asyncpp/0.2.3**

Reason for addition:
- I'm the author of the library
- I'd like to make it more accessible to people

Why is this library useful for people:
- Not many options currently for coroutine libraries
- Makes coroutines practical for developers

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

RubenRBS commented:
Hi @petiaccja thanks a lot for taking the time to create the recipe, we really appreciate it. I'll help you move this along and will be pushing some minor changes in a bit directly to your branch if that's ok with you :)

petiaccja commented:
> Hi @petiaccja thanks a lot for taking the time to create the recipe, we really appreciate it. I'll help you move this along and will be pushing some minor changes in a bit directly to your branch if that's ok with you :)

Hey @RubenRBS, thank you for helping out with the PR, and I hope you don't mind that I made a draft so that I could leverage the CI. Feel free to push some changes :)

RubenRBS reviewed (commented):
@petiaccja check my comments, and let me know your suggestions, thanks!

petiaccja commented on recipes/asyncpp/all/conanfile.py:
```diff
@@ -0,0 +1,104 @@
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "13",
+            "clang": "14",
```
I've pushed a commit to update the version strings to the ones that I use on the CI.

It's possible that it works for earlier versions of the compilers as well, but I don't have a setup for that right now, so maybe I can update the compiler version table in a future PR.

In the case of MSVC, `"193"` means anything above 19.30, right? Can I also specify `"1932"` for example? I think 19.30 should be fine in all cases though.

RubenRBS replied:
Hi sorry I missed this question: In CCI, we expect users to have the latest point release of msvc available, so checking for "1932" would not be needed :)

petiaccja commented on recipes/asyncpp/all/conanfile.py:
```diff
@@ -0,0 +1,104 @@
+    def package_info(self):
+        self.cpp_info.libs = ["asyncpp"]
+        self.cpp_info.set_property("cmake_file_name", "asyncpp")
+        self.cpp_info.set_property("cmake_target_name", "asyncpp")
```
I think the best would be to stick with conan's conventions, so that the package name is consistent with most other libraries out there.

The examples and benchmarks are not so important, as they are mostly used for developing the library, and can be changed easily without modifying the conan recipe.

petiaccja replied:
I'm not really up to date on conan's conventions, so I'll leave the choice up to you :)

petiaccja replied:
Oh, I see. In that case I think I'd stick with `asyncpp::asyncpp`, that would make it more future proof if I want to add something like `asyncpp::network`.

petiaccja commented on recipes/asyncpp/all/conanfile.py:
```diff
@@ -0,0 +1,104 @@
+        cmake = CMake(self)
+        cmake.install()
+
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Good catch! I've deleted them and checked `~\.conan2\p\b\async<hash>\p`, and there doesn't seem to be any clutter.

I think these were needed when I haven't had a proper install script in the `CMakeLists.txt` of asyncpp, and I accidentally left them in.

petiaccja commented on recipes/asyncpp/all/conanfile.py:
```diff
@@ -0,0 +1,104 @@
+                f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+            )
+
+        if self.settings.compiler == "apple-clang":
```
Yes, that makes sense, as I haven't tested it on Apple-Clang. It would be possible to replace `jthread` with the good old `thread`, however, I'd like to keep complexity to a minimum by using `jthread` as multi-threaded code is usually bad enough. I think while asyncpp matures a bit, Apple-Clang might catch up on the features. Do you think it's fine to not support Apple-Clang at the moment?

RubenRBS replied:
Of course! Let's keep this line then and whenever newer apple-clang versions support it, we can come back and allow those verisons :)

RubenRBS commented:
Weird, this now builds locally for me

uilianries commented:
> Weird, this now builds locally for me

@RubenRBS I can build locally, I have Windows™ with amd64 here.

uilianries commented on recipes/asyncpp/all/conanfile.py:
```diff
@@ -0,0 +1,102 @@
+
+        if self.settings.compiler == "apple-clang":
+            # As per https://en.cppreference.com/w/cpp/compiler_support, apple-clang 14.0.0 partially supports C++20 but not jthread
+            raise ConanInvalidConfiguration(f"{self.ref} does not support apple-clang compiler, as it lacks jthread support.")
```
```suggestion
            raise ConanInvalidConfiguration(f"{self.ref} does not support apple-clang compiler, as it lacks jthread support.")
            
        if is_msvc(self) and self.options.shared:
            # INFO: asyncpp does not contain .def neither provides .lib when building shared
            raise ConanInvalidConfiguration(f"{self.ref} does not support -o {self.ref}:shared=True with compiler=msvc")
```

The current error is because the project only provides a DLL file, but does not expose symbols in a .lib: https://c3i.jfrog.io/c3i/misc-v2/logs/pr/23974/8-windows-msvc/asyncpp/0.2.3//4dfc129aabdd7b56484b8f1e47f0337c1164e51e-build.txt

```
asyncpp/0.2.3: RUN: cmake --install "C:\J\workspace\prod-v2\bsr\39245\fbfcf\p\b\async8b8db22574e05\b\build" --config Release --prefix "C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p"
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/lib/asyncpp.dll
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/CMakeLists.txt
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/concepts.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_collection.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_deque.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_item.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_stack.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/event.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/generator.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/join.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/lock.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/memory
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/memory/rc_ptr.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/mutex.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/promise.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/scheduler.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/semaphore.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/shared_mutex.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/sleep.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/stream.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/task.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing/interleaver.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing/suspension_point.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading/cache.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading/spinlock.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/thread_pool.hpp
```

People would need to use WINAPI LoadLibrary and load symbols manually ... 

As you are the author, you could adjust the project to export symbols:

- Adding the CMake definition WINDOWS_EXPORT_ALL_SYMBOLS. It works most of cases, but is fragile, will expose all methods, including internals. It does not export static public variables, it will result in runtime errors
- Use __declspec to export only what you want: https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll?view=msvc-170
- Use CMake GenerateExportHeader to automatically export functions: https://www.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/

petiaccja replied:
Hey @RubenRBS and @uilianries, thanks for checking out this issue. I won't have time in the upcoming week, but I'm going to modify the library so that the symbols are exported properly and get back to you.

petiaccja commented:
Hey guys, I've added support for building the library in shared mode and I've updated the recipe to use the new library version.

Things look green now, however, scrolling through the logs I see that the only configuration that gets built is MSVC 193. All the other configurations on Conan's CI use compilers that are too old for this library, so they are getting discarded as `INVALID_CONFIGURATION`, meaning the actual coverage of configurations by the CI is really low. Is there any way to include newer compilers on the CI on a case-by-case basis, or any plans to keep more recent ones on the CI by default?

RubenRBS commented:
Writing here to let you know that I've read this, will answer tomorrow once I get a chance to get in front of a computer :)

petiaccja commented:
No rush :)

RubenRBS commented:
> Things look green now, however, scrolling through the logs I see that the only configuration that gets built is MSVC 193. All the other configurations on Conan's CI use compilers that are too old for this library, so they are getting discarded as INVALID_CONFIGURATION, meaning the actual coverage of configurations by the CI is really low. Is there any way to include newer compilers on the CI on a case-by-case basis, or any plans to keep more recent ones on the CI by default?

Once we stop supporting Conan 1, we plan on adding new configurations for newer compilers as part of the Conan 2 build matrix. For now this compiles in Windows which is enough for us to approve, but I'd appreciate it if you could provide compilation logs for gcc/linux showing that it does in fact compile for that configuration, it will help tremendously with future traceability :)

petiaccja commented:
> Once we stop supporting Conan 1, we plan on adding new configurations for newer compilers as part of the Conan 2 build matrix. For now this compiles in Windows which is enough for us to approve, but I'd appreciate it if you could provide compilation logs for gcc/linux showing that it does in fact compile for that configuration, it will help tremendously with future traceability :)

Thanks for the information :) I've quickly checked it under Linux with both GCC 13 and Clang 18 with both shared and static linking, and it seems to work:

[asyncpp_conan_clang_log.txt](https://github.com/user-attachments/files/15686011/asyncpp_conan_clang_log.txt)
[asyncpp_conan_clang_sh_log.txt](https://github.com/user-attachments/files/15686012/asyncpp_conan_clang_sh_log.txt)
[asyncpp_conan_gcc_log.txt](https://github.com/user-attachments/files/15686015/asyncpp_conan_gcc_log.txt)
[asyncpp_conan_gcc_sh_log.txt](https://github.com/user-attachments/files/15686016/asyncpp_conan_gcc_sh_log.txt)

PS: I only had Clang 18 around, but the CI runs on 17 so I expect that works with conan too, as specified in the recipe.

perseoGI reviewed (approved):
LGTM, we need more easy coroutine libs! Thanks a lot 💯

RubenRBS commented:
> Thanks for the information :) I've quickly checked it under Linux with both GCC 13 and Clang 18 with both shared and static linking, and it seems to work:

Thanks a lot for the detailed compilation logs, really appreciated!

petiaccja commented:
Thanks a lot for supervising and reviewing this PR :)
